go/vt/vttablet/tabletserver/query_executor.go

```diff
@@ -387,14 +390,31 @@ func (qre *QueryExecutor) checkPermissions() error {
 		remoteAddr = ci.RemoteAddr()
 		username = ci.Username()
 	}
-	action, desc := qre.plan.Rules.GetAction(remoteAddr, username, qre.bindVars, qre.marginComments)
-	switch action {
-	case rules.QRFail:
-		return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "disallowed due to rule: %s", desc)
-	case rules.QRFailRetry:
-		return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "disallowed due to rule: %s", desc)
-	}
 
+	ctx, cancel := context.WithTimeout(qre.ctx, maxQueryBufferDuration)
+	defer cancel()
+	for goodToProceed := false; !goodToProceed; {
```

Comment:
loop while buffering is active. If there's no buffering rule, this loop only executes once.